--- ../src-base/minecraft/ic2/core/energy/leg/EnergyCalculatorLeg.java
+++ ../src-work/minecraft/ic2/core/energy/leg/EnergyCalculatorLeg.java
@@ -145,10 +145,10 @@
             return list;
          } else {
             List<EnergyPath> arraylist = new ArrayList();
-            Iterator iterator = data.energySourceToEnergyPathMap.values().iterator();
+            Iterator<List<EnergyPath>> iterator = data.energySourceToEnergyPathMap.values().iterator();
 
             while(iterator.hasNext()) {
-               for(EnergyPath energypath : (List)iterator.next()) {
+               for(EnergyPath energypath : iterator.next()) {
                   if (node.getType() == NodeType.Sink) {
                      if (energypath.target == node) {
                         arraylist.add(energypath);
