--- ../src-base/minecraft/ic2/core/block/machine/tileentity/TileEntityFluidBottler.java
+++ ../src-work/minecraft/ic2/core/block/machine/tileentity/TileEntityFluidBottler.java
@@ -59,7 +59,7 @@
       this.outputSlot.add(processResult);
    }
 
-   public MachineRecipeResult<Void, Object, Object> getOutput() {
+   public MachineRecipeResult getOutput() {
       MachineRecipeResult<Void, IEmptyFluidContainerRecipeManager.Output, ItemStack> machinereciperesult = Recipes.emptyFluidContainer.apply(this.drainInputSlot.get(), this.fluidTank.getFluid() == null ? null : this.fluidTank.getFluid().getFluid(), FluidContainerOutputMode.EmptyFullToOutput, false);
       if (machinereciperesult != null && (machinereciperesult.getOutput()).fluid.amount <= this.fluidTank.getCapacity() - this.fluidTank.getFluidAmount() && this.outputSlot.canAdd((machinereciperesult.getOutput()).container)) {
          return machinereciperesult;
