--- a/ic2/core/block/machine/container/ContainerBatchCrafter.java
+++ b/ic2/core/block/machine/container/ContainerBatchCrafter.java
@@ -4,6 +4,7 @@
 import gnu.trove.iterator.TIntIterator;
 import gnu.trove.map.TIntIntMap;
 import gnu.trove.map.hash.TIntIntHashMap;
+import ic2.api.upgrade.IUpgradeItem;
 import ic2.core.block.machine.tileentity.TileEntityBatchCrafter;
 import ic2.core.slot.SlotHologramSlot;
 import ic2.core.slot.SlotInvSlot;
@@ -11,6 +12,8 @@
 import ic2.core.util.Tuple;
 import java.util.Collections;
 import java.util.List;
+import java.util.Stack;
+
 import net.minecraft.entity.player.EntityPlayer;
 import net.minecraft.inventory.Slot;
 import net.minecraft.item.ItemStack;
@@ -45,12 +48,23 @@
       }
 
       for (int slot = 0; slot < 4; slot++) {
-         this.addSlotToContainer(new SlotInvSlot(tileEntity.upgradeSlot, slot, 152, 8 + slot * 18));
+         // Ingredients: 0-8 // Upgrades: 9-12
+         this.indexToSlot.put(slot+9, this.addSlotToContainer(new SlotInvSlot(tileEntity.upgradeSlot, slot, 152, 8 + slot * 18)).slotNumber);
       }
    }
 
    @Override
    protected ItemStack handlePlayerSlotShiftClick(EntityPlayer player, ItemStack sourceItemStack) {
+      if (sourceItemStack.getItem() instanceof IUpgradeItem && this.base.upgradeSlot.accepts(sourceItemStack)) {
+         for (int i = 0; i < 4; i++) {
+            Slot slot = this.inventorySlots.get(this.indexToSlot.get(i+9));
+            if (slot.isItemValid(sourceItemStack)) {
+               sourceItemStack = this.transfer(sourceItemStack, slot);
+               if (StackUtil.isEmpty(sourceItemStack)) return StackUtil.emptyStack;
+            }
+         }
+      }
+
       Tuple.T2<List<ItemStack>, ? extends TIntCollection> changes = StackUtil.balanceStacks(
          this.base.ingredients, this.base.acceptPredicate, StackUtil.getSlotsFromInv(this.base.ingredients), Collections.singleton(sourceItemStack)
       );
@@ -58,10 +72,14 @@
 
       while (iter.hasNext()) {
          int currentSlot = iter.next();
-         ((Slot)this.inventorySlots.get(this.indexToSlot.get(currentSlot))).onSlotChanged();
+         this.inventorySlots.get(this.indexToSlot.get(currentSlot)).onSlotChanged();
       }
+      if (changes.a.isEmpty()) return StackUtil.emptyStack;
+      sourceItemStack = changes.a.get(0);
+      if (this.base.dischargeSlot.accepts(sourceItemStack))
+         sourceItemStack = this.transfer(sourceItemStack, this.inventorySlots.get(36));
 
-      return changes.a.isEmpty() ? StackUtil.emptyStack : changes.a.get(0);
+      return sourceItemStack;
    }
 
    @Override
