--- a/ic2/core/item/armor/ItemArmorQuantumSuit.java
+++ b/ic2/core/item/armor/ItemArmorQuantumSuit.java
@@ -215,7 +215,7 @@
                IC2.achievements.issueAchievement(player, "starveWithQHelmet");
             }
 
-            for (PotionEffect effect : new LinkedList(player.getActivePotionEffects())) {
+            for (PotionEffect effect : new LinkedList<>(player.getActivePotionEffects())) {
                Potion potion = effect.getPotion();
                Integer cost = potionRemovalCost.get(potion);
                if (cost != null) {
