--- ../src-base/minecraft/ic2/core/item/armor/ItemArmorQuantumSuit.java
+++ ../src-work/minecraft/ic2/core/item/armor/ItemArmorQuantumSuit.java
@@ -278,6 +278,23 @@
          IC2.platform.profilerStartSection("QuantumBodyarmor");
          player.extinguish();
          IC2.platform.profilerEndSection();
+
+         if (player.isSneaking() && IC2.keyboard.isModeSwitchKeyDown(player) && b0 == 0) {
+            b0 = 10;
+            boolean disabled = !nbttagcompound.getBoolean("jetpackDisabled");
+            if (IC2.platform.isSimulating()) {
+               nbttagcompound.setBoolean("jetpackDisabled", disabled);
+               IC2.platform.messagePlayer(player, "Jetpack "+(disabled ? "disabled" : "enabled"));
+               if (disabled) nbttagcompound.setBoolean("hoverMode", false);
+            } else if (disabled) {
+               IC2.audioManager.removeSources(player);
+            }
+         }
+
+         if (IC2.platform.isSimulating() && b0 > 0) {
+            --b0;
+            nbttagcompound.setByte("toggleTimer", b0);
+         }
          break;
       case LEGS:
          IC2.platform.profilerStartSection("QuantumLeggings");
@@ -375,7 +392,7 @@
    }
 
    public boolean isJetpackActive(ItemStack stack) {
-      return true;
+      return !StackUtil.getOrCreateNbtData(stack).getBoolean("jetpackDisabled");
    }
 
    public float getHoverMultiplier(ItemStack stack, boolean upwards) {
