--- ../src-base/minecraft/ic2/core/block/steam/BlockRefractoryBricks.java
+++ ../src-work/minecraft/ic2/core/block/steam/BlockRefractoryBricks.java
@@ -1,6 +1,7 @@
 package ic2.core.block.steam;
 
 import ic2.core.block.BlockBase;
+import ic2.core.profile.NotClassic;
 import ic2.core.ref.BlockName;
 import net.minecraft.block.material.Material;
 import net.minecraft.block.properties.IProperty;
@@ -9,6 +10,7 @@
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.IBlockAccess;
 
+@NotClassic
 public class BlockRefractoryBricks extends BlockBase {
    public BlockRefractoryBricks() {
       super(BlockName.refractory_bricks, Material.ROCK);
