--- ../src-base/minecraft/ic2/core/block/steam/TileEntityCokeKilnGrate.java
+++ ../src-work/minecraft/ic2/core/block/steam/TileEntityCokeKilnGrate.java
@@ -10,12 +10,15 @@
 import ic2.core.gui.dynamic.GuiParser;
 import ic2.core.network.GuiSynced;
 import java.util.Collections;
+
+import ic2.core.profile.NotClassic;
 import net.minecraft.client.gui.GuiScreen;
 import net.minecraft.entity.player.EntityPlayer;
 import net.minecraft.util.EnumFacing;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
 
+@NotClassic
 public class TileEntityCokeKilnGrate extends TileEntityInventory implements IHasGui {
    protected final Fluids fluids = (Fluids)this.addComponent(new Fluids(this));
    @GuiSynced
