--- ../src-base/minecraft/ic2/core/util/ReflectionUtil.java
+++ ../src-work/minecraft/ic2/core/util/ReflectionUtil.java
@@ -126,18 +126,17 @@
 
    public static void setValue(Object object, Field field, Object value) {
       if (field.getType().isEnum() && value instanceof Integer) {
-         value = field.getType().getEnumConstants()[((Integer)value).intValue()];
+         value = field.getType().getEnumConstants()[(Integer)value];
       }
-
       try {
-         Object object = field.get(object);
-         if (!DataEncoder.copyValue(value, object)) {
+         Object oldValue = field.get(object);
+         if (!DataEncoder.copyValue(value, oldValue)) {
             field.set(object, value);
          }
-
-      } catch (Exception exception) {
-         throw new RuntimeException("can't set field " + field.getName() + " in " + object + " to " + value, exception);
       }
+      catch (Exception e) {
+         throw new RuntimeException("can't set field " + field.getName() + " in " + object + " to " + value, e);
+      }
    }
 
    public static boolean setValueRecursive(Object object, String fieldName, Object value) {
