--- a/ic2/core/item/tool/ItemElectricToolChainsaw.java
+++ b/ic2/core/item/tool/ItemElectricToolChainsaw.java
@@ -124,8 +124,8 @@
          if (block instanceof IShearable) {
             IShearable ishearable = (IShearable)block;
             if (ishearable.isShearable(itemstack, world, pos) && ElectricItem.manager.use(itemstack, this.operationEnergyCost, player)) {
-               for(ItemStack itemstack : ishearable.onSheared(itemstack, world, pos, EnchantmentHelper.getEnchantmentLevel(Enchantments.FORTUNE, itemstack))) {
-                  StackUtil.dropAsEntity(world, pos, itemstack);
+               for(ItemStack stack : ishearable.onSheared(itemstack, world, pos, EnchantmentHelper.getEnchantmentLevel(Enchantments.FORTUNE, itemstack))) {
+                  StackUtil.dropAsEntity(world, pos, stack);
                }
 
                player.addStat(StatList.getBlockStats(block), 1);
