--- a/ic2/core/profile/ProfileManager.java
+++ b/ic2/core/profile/ProfileManager.java
@@ -8,11 +8,8 @@
 import ic2.core.util.Config;
 import ic2.core.util.LogCategory;
 import ic2.core.util.ReflectionUtil;
-import java.io.ByteArrayInputStream;
-import java.io.File;
-import java.io.IOException;
-import java.io.InputStream;
-import java.io.SequenceInputStream;
+
+import java.io.*;
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.List;
@@ -61,7 +58,7 @@
          if (root.exists()) {
             for (File file : root.listFiles()) {
                if (file.isDirectory()) {
-                  for (File profile : file.listFiles(new NameFileFilter("profile.xml"))) {
+                  for (File profile : file.listFiles((FileFilter) new NameFileFilter("profile.xml"))) {
                      try {
                         Profile p = ProfileParser.parse(new ProfileRoot(profile.getParentFile()));
                         if (!registerProfile(p)) {
