--- a/ic2/core/block/machine/tileentity/TileEntityBatchCrafter.java
+++ b/ic2/core/block/machine/tileentity/TileEntityBatchCrafter.java
@@ -297,7 +297,7 @@
    }
 
    protected void doCrafting() {
-      for(int j = 0; j < this.operationsPerTick; ++j) {
+      for(int j = 0; j < this.operationsPerTick && canCraft(); ++j) {
          List<ItemStack> list = Collections.<ItemStack>singletonList(this.recipeOutput);
 
          for(ItemStack itemstack : this.upgradeSlot) {
