--- a/ic2/core/uu/RecipeResolver.java
+++ b/ic2/core/uu/RecipeResolver.java
@@ -27,19 +27,15 @@
    }
 
    private static List<List<LeanItemStack>> toDoubleStackList(List<Ingredient> list) {
-      List<List<LeanItemStack>> list = new ArrayList<List<LeanItemStack>>(list.size());
-
-      for(Ingredient ingredient : list) {
-         ItemStack[] aitemstack = ingredient.getMatchingStacks();
-         List<LeanItemStack> list1 = new ArrayList<LeanItemStack>(aitemstack.length);
-
-         for(ItemStack itemstack : aitemstack) {
-            list1.add(new LeanItemStack(itemstack));
+      ArrayList<List<LeanItemStack>> ret = new ArrayList<List<LeanItemStack>>(list.size());
+      for (Ingredient ingredient : list) {
+         ItemStack[] arr = ingredient.getMatchingStacks();
+         ArrayList<LeanItemStack> toAdd = new ArrayList<LeanItemStack>(arr.length);
+         for (ItemStack stack : arr) {
+            toAdd.add(new LeanItemStack(stack));
          }
-
-         list.add(list1);
+         ret.add(toAdd);
       }
-
-      return list;
+      return ret;
    }
 }
