--- a/ic2/core/item/tool/ItemDebug.java
+++ b/ic2/core/item/tool/ItemDebug.java
@@ -369,7 +369,7 @@
                } else if (object instanceof Map) {
                   ps.println("  values (" + ((Map)object).size() + "):");
 
-                  for(Entry<?, ?> entry : ((Map)object).entrySet()) {
+                  for(Entry<?, ?> entry : ((Map<?, ?>)object).entrySet()) {
                      ps.print("    " + entry.getKey() + ": ");
                      dumpValueString(entry.getValue(), field, "      ", ps);
                   }
