--- ../src-base/minecraft/ic2/core/item/armor/ItemArmorUtility.java
+++ ../src-work/minecraft/ic2/core/item/armor/ItemArmorUtility.java
@@ -12,7 +12,7 @@
 
 public class ItemArmorUtility extends ItemArmorIC2 implements ISpecialArmor {
    public ItemArmorUtility(ItemName name, String armorName, EntityEquipmentSlot type) {
-      super(name, ArmorMaterial.DIAMOND, armorName, type, (Object)null);
+      super(name, ArmorMaterial.DIAMOND, armorName, type, null);
    }
 
    public int getItemEnchantability() {
