--- a/ic2/core/energy/grid/Tile.java
+++ b/ic2/core/energy/grid/Tile.java
@@ -52,24 +52,21 @@
    }
 
    boolean removeExtraNode(Node node) {
-      boolean flag = false;
+      boolean canBeRemoved = false;
       if (node.isExtraNode()) {
-         flag = true;
+         canBeRemoved = true;
       } else {
-         for(Node node : this.nodes) {
-            if (node != node && node.nodeType == node.nodeType && node.isExtraNode()) {
-               node.setExtraNode(false);
-               flag = true;
-               break;
-            }
+         for (Node otherNode : this.nodes) {
+            if (otherNode == node || otherNode.nodeType != node.nodeType || !otherNode.isExtraNode()) continue;
+            otherNode.setExtraNode(false);
+            canBeRemoved = true;
+            break;
          }
       }
-
-      if (flag) {
+      if (canBeRemoved) {
          this.nodes.remove(node);
       }
-
-      return flag;
+      return canBeRemoved;
    }
 
    public Collection<IEnergyTile> getSubTiles() {
