--- a/ic2/core/gui/dynamic/TextProvider.java
+++ b/ic2/core/gui/dynamic/TextProvider.java
@@ -56,7 +56,7 @@
             list = new ArrayList<TextProvider.AbstractTextProvider>();
          } else if (c0 == '{' && c1 == ',') {
             finish(stringbuilder1, list);
-            list.add((Object)null);
+            list.add(null);
          } else if (c0 == '{' && c1 == '}') {
             finish(stringbuilder1, list);
             TextProvider.AbstractTextProvider textprovider$abstracttextprovider2 = null;
@@ -150,7 +150,7 @@
       if (s == null) {
          return new TextProvider.TokenResolverDynamic(token);
       } else {
-         String s1 = resolveToken(s, baseClass, (Object)null, emptyTokens());
+         String s1 = resolveToken(s, baseClass, null, emptyTokens());
          return (TextProvider.AbstractTextProvider)(s1 != null ? new TextProvider.Constant(s1) : new TextProvider.TokenResolverStatic(s));
       }
    }
@@ -384,7 +384,7 @@
 
       public String getConstant(Class<?> baseClass) {
          String s = this.token.getConstant(baseClass);
-         return s == null ? null : TextProvider.resolveToken(s, baseClass, (Object)null, Collections.emptyMap());
+         return s == null ? null : TextProvider.resolveToken(s, baseClass, null, Collections.emptyMap());
       }
    }
 
@@ -400,7 +400,7 @@
       }
 
       public String getConstant(Class<?> baseClass) {
-         return TextProvider.resolveToken(this.token, baseClass, (Object)null, Collections.emptyMap());
+         return TextProvider.resolveToken(this.token, baseClass, null, Collections.emptyMap());
       }
    }
 
