--- a/ic2/core/block/state/MaterialProperty.java
+++ b/ic2/core/block/state/MaterialProperty.java
@@ -105,7 +105,7 @@
          try {
             for(Field field : Material.class.getFields()) {
                if (field.getType() == Material.class) {
-                  Material material = (Material)field.get((Object)null);
+                  Material material = (Material)field.get(null);
                   MATERIAL_TO_WRAP.put(material, new MaterialProperty.WrappedMaterial(material, field.getName()));
                }
             }
