--- a/ic2/core/block/invslot/InvSlotDischarge.java
+++ b/ic2/core/block/invslot/InvSlotDischarge.java
@@ -9,6 +9,7 @@
 import net.minecraft.item.ItemStack;
 
 public class InvSlotDischarge extends InvSlot implements IDischargingSlot {
+   public int baseTier;
    public int tier;
    public boolean allowRedstoneDust = true;
 
@@ -23,6 +24,7 @@
    public InvSlotDischarge(IInventorySlotHolder<?> base, InvSlot.Access access, int tier, boolean allowRedstoneDust, InvSlot.InvSide preferredSide) {
       super(base, "discharge", access, 1, preferredSide);
       this.tier = tier;
+      this.baseTier = tier;
       this.allowRedstoneDust = allowRedstoneDust;
    }
 
