--- a/ic2/core/command/CommandIc2.java
+++ b/ic2/core/command/CommandIc2.java
@@ -464,7 +464,7 @@
             if (this.pattern.matcher(regName).matches()) {
                if (this.meta == null) {
                   if (item instanceof IMultiItem) {
-                     for (ItemStack stack : ((IMultiItem)item).getAllStacks()) {
+                     for (ItemStack stack : ((IMultiItem<?>)item).getAllStacks()) {
                         assert stack != null : item + " produced a null stack in getAllStacks()";
 
                         this.dump(stack, regName);
