--- a/ic2/core/ref/MetaTeBlockProperty.java
+++ b/ic2/core/ref/MetaTeBlockProperty.java
@@ -38,7 +38,7 @@
                private final List<MetaTeBlockProperty.MetaTePair> teBlockMap = (List<MetaTeBlockProperty.MetaTePair>)MetaTeBlockProperty.resourceToTeBlock
                   .get(identifier)
                   .b;
-               private final int allTeBlockSize = <unrepresentable>.super.teBlockMap.size();
+               private final int allTeBlockSize = this.teBlockMap.size();
 
                @Override
                public boolean hasNext() {
@@ -150,7 +150,7 @@
          if (type.getKey() != invalid.teBlock.getIdentifier()) {
             Tuple.T2 var10 = type.getValue();
             Integer var11 = (Integer)var10.a;
-            Object var12 = var10.a = (TA)(Integer)var10.a + 1;
+            Object var12 = var10.a = (Integer)var10.a + 1;
             ((List)type.getValue().b).add(invalidStates);
          }
       }
