--- a/ic2/core/network/GrowingBuffer.java
+++ b/ic2/core/network/GrowingBuffer.java
@@ -213,18 +213,15 @@
       if (i < 0) {
          throw new IllegalArgumentException("only positive numbers are supported");
       } else {
-         while(true) {
-            int i = i & 127;
+         do {
+            int part = i & 127;
             i >>>= 7;
             if (i != 0) {
-               i |= 128;
+               part |= 128;
             }
 
-            this.writeByte(i);
-            if (i == 0) {
-               break;
-            }
-         }
+            this.writeByte(part);
+         } while(i != 0);
 
       }
    }
