--- a/ic2/core/init/BlocksItems.java
+++ b/ic2/core/init/BlocksItems.java
@@ -281,8 +281,9 @@
             return !IC2.version.isClassic();
          }
       };
-      new ItemBattery(ItemName.energy_crystal, 1000000.0D, 2048.0D, 3);
-      (new ItemBattery(ItemName.lapotron_crystal, 1.0E7D, 8092.0D, 4)).setRarity(EnumRarity.UNCOMMON);
+      boolean classic = IC2.version.isClassic();
+      new ItemBattery(ItemName.energy_crystal, classic ? 100000 : 1000000, classic ? 250 : 2048, classic ? 2 : 3);
+      (new ItemBattery(ItemName.lapotron_crystal, classic ? 1000000 : 1.0E7D, classic ? 600 : 8092.0D, classic ? 3 : 4)).setRarity(EnumRarity.UNCOMMON);
       new ItemBatterySU(ItemName.single_use_battery, 1200, 1);
       new ItemBatteryChargeHotbar(ItemName.charging_re_battery, 40000.0D, 128.0D, 1);
       new ItemBatteryChargeHotbar(ItemName.advanced_charging_re_battery, 400000.0D, 1024.0D, 2);
