--- ../src-base/minecraft/ic2/core/block/machine/EmptyFluidContainerRecipeManager.java
+++ ../src-work/minecraft/ic2/core/block/machine/EmptyFluidContainerRecipeManager.java
@@ -30,7 +30,7 @@
             return null;
          } else {
             Collection<ItemStack> collection = StackUtil.isEmpty(liquidutil$fluidoperationresult.extraOutput) ? Collections.emptyList() : Collections.singletonList(liquidutil$fluidoperationresult.extraOutput);
-            return (new MachineRecipe((Object)null, new IEmptyFluidContainerRecipeManager.Output(collection, liquidutil$fluidoperationresult.fluidChange))).getResult(liquidutil$fluidoperationresult.inPlaceOutput);
+            return (new MachineRecipe(null, new IEmptyFluidContainerRecipeManager.Output(collection, liquidutil$fluidoperationresult.fluidChange))).getResult(liquidutil$fluidoperationresult.inPlaceOutput);
          }
       }
    }
