--- a/ic2/core/block/machine/EmptyFluidContainerRecipeManager.java
+++ b/ic2/core/block/machine/EmptyFluidContainerRecipeManager.java
@@ -33,7 +33,7 @@
             return null;
          } else {
             Collection<ItemStack> output = StackUtil.isEmpty(result.extraOutput) ? Collections.emptyList() : Collections.singletonList(result.extraOutput);
-            return new MachineRecipe<>(null, new IEmptyFluidContainerRecipeManager.Output(output, result.fluidChange)).getResult(result.inPlaceOutput);
+            return new MachineRecipe(null, new IEmptyFluidContainerRecipeManager.Output(output, result.fluidChange)).getResult(result.inPlaceOutput);
          }
       }
    }
