--- a/ic2/core/item/tool/ItemToolWrenchNew.java
+++ b/ic2/core/item/tool/ItemToolWrenchNew.java
@@ -112,7 +112,7 @@
          return false;
       } else {
          TileEntity tileentity = world.getTileEntity(pos);
-         return tileentity instanceof IPipe || Arrays.stream((Object[])EnumFacing.VALUES).anyMatch((face) -> {
+         return tileentity instanceof IPipe || Arrays.stream(EnumFacing.VALUES).anyMatch((face) -> {
             return ((IWrenchable)block).canSetFacing(world, pos, face, player);
          });
       }
