--- a/ic2/core/gui/dynamic/DynamicContainer.java
+++ b/ic2/core/gui/dynamic/DynamicContainer.java
@@ -74,49 +74,49 @@
                   throw new RuntimeException("invalid base " + this.base + " for slot elements");
                }
 
-               GuiParser.SlotGridNode node = (GuiParser.SlotGridNode)rawNode;
-               InvSlot slot = ((TileEntityInventory)this.base).getInvSlot(node.name);
-               if (slot == null) {
-                  throw new RuntimeException("invalid invslot name " + node.name + " for base " + this.base);
+               GuiParser.SlotGridNode gridNode = (GuiParser.SlotGridNode)rawNode;
+               InvSlot gridSlot = ((TileEntityInventory)this.base).getInvSlot(gridNode.name);
+               if (gridSlot == null) {
+                  throw new RuntimeException("invalid invslot name " + gridNode.name + " for base " + this.base);
                }
 
-               int size = slot.size();
-               if (size > node.offset) {
-                  int x0 = node.x + (node.style.width - 16) / 2;
-                  int y0 = node.y + (node.style.height - 16) / 2;
-                  GuiParser.SlotGridNode.SlotGridDimension dim = node.getDimension(size);
+               int size = gridSlot.size();
+               if (size > gridNode.offset) {
+                  int x0 = gridNode.x + (gridNode.style.width - 16) / 2;
+                  int y0 = gridNode.y + (gridNode.style.height - 16) / 2;
+                  GuiParser.SlotGridNode.SlotGridDimension dim = gridNode.getDimension(size);
                   int rows = dim.rows;
                   int cols = dim.cols;
-                  int width = node.style.width + node.spacing;
-                  int height = node.style.height + node.spacing;
-                  int idx = node.offset;
-                  if (!node.vertical) {
-                     int y = y0;
+                  int gridWidth = gridNode.style.width + gridNode.spacing;
+                  int gridHeight = gridNode.style.height + gridNode.spacing;
+                  int idx = gridNode.offset;
+                  if (!gridNode.vertical) {
+                     int gridY = y0;
 
                      for (int row = 0; row < rows && idx < size; row++) {
-                        int x = x0;
+                        int gridX = x0;
 
                         for (int col = 0; col < cols && idx < size; col++) {
-                           this.addSlotToContainer(new SlotInvSlot(slot, idx, x, y));
+                           this.addSlotToContainer(new SlotInvSlot(gridSlot, idx, gridX, gridY));
                            idx++;
-                           x += width;
+                           gridX += gridWidth;
                         }
 
-                        y += height;
+                        gridY += gridHeight;
                      }
                   } else {
-                     int x = x0;
+                     int gridX = x0;
 
                      for (int col = 0; col < cols && idx < size; col++) {
-                        int y = y0;
+                        int gridY = y0;
 
                         for (int row = 0; row < rows && idx < size; row++) {
-                           this.addSlotToContainer(new SlotInvSlot(slot, idx, x, y));
+                           this.addSlotToContainer(new SlotInvSlot(gridSlot, idx, gridX, gridY));
                            idx++;
-                           y += height;
+                           gridY += gridHeight;
                         }
 
-                        x += width;
+                        gridX += gridWidth;
                      }
                   }
                }
@@ -126,12 +126,12 @@
                   throw new RuntimeException("Invalid base " + this.base + " for holographic slot elements");
                }
 
-               GuiParser.SlotHologramNode node = (GuiParser.SlotHologramNode)rawNode;
-               int x = node.x + (node.style.width - 16) / 2;
-               int y = node.y + (node.style.height - 16) / 2;
+               GuiParser.SlotHologramNode hologramNode = (GuiParser.SlotHologramNode)rawNode;
+               int hologramX = hologramNode.x + (hologramNode.style.width - 16) / 2;
+               int hologramY = hologramNode.y + (hologramNode.style.height - 16) / 2;
                this.addSlotToContainer(
                   new SlotHologramSlot(
-                     ((IHolographicSlotProvider)this.base).getStacksForName(node.name), node.index, x, y, node.stackSizeLimit, this.getCallback()
+                     ((IHolographicSlotProvider)this.base).getStacksForName(hologramNode.name), hologramNode.index, hologramX, hologramY, hologramNode.stackSizeLimit, this.getCallback()
                   )
                );
             case gui:
