--- ../src-base/minecraft/ic2/core/crop/CropModel.java
+++ ../src-work/minecraft/ic2/core/crop/CropModel.java
@@ -53,12 +53,12 @@
 
          for(CropCard cropcard : Crops.instance.getCrops()) {
             for(ResourceLocation resourcelocation : cropcard.getTexturesLocation()) {
-               textures.put(resourcelocation, (Object)null);
+               textures.put(resourcelocation, null);
             }
          }
 
-         textures.put(STICK, (Object)null);
-         textures.put(UPGRADED_STICK, (Object)null);
+         textures.put(STICK, null);
+         textures.put(UPGRADED_STICK, null);
       }
 
       return textures.keySet();
