--- a/ic2/core/block/invslot/InvSlotUpgrade.java
+++ b/ic2/core/block/invslot/InvSlotUpgrade.java
@@ -18,6 +18,8 @@
 import java.util.List;
 import net.minecraft.item.Item;
 import net.minecraft.item.ItemStack;
+import ic2.api.upgrade.UpgradableProperty;
+import ic2.core.block.machine.tileentity.TileEntityElectricMachine;
 
 public class InvSlotUpgrade extends InvSlot {
    private static final int maxStackSize = 64;
@@ -90,6 +92,11 @@
                }
             }
          }
+      }
+
+      if (iupgradableblock instanceof TileEntityElectricMachine && iupgradableblock.getUpgradableProperties().contains(UpgradableProperty.Transformer)) {
+         TileEntityElectricMachine te = ((TileEntityElectricMachine) iupgradableblock);
+         te.dischargeSlot.setTier(te.dischargeSlot.baseTier + extraTier);
       }
 
       for(TileEntityComponent tileentitycomponent : this.base.getParent().getComponents()) {
