--- a/ic2/core/WorldData.java
+++ b/ic2/core/WorldData.java
@@ -13,6 +13,8 @@
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.ConcurrentLinkedQueue;
 import java.util.concurrent.ConcurrentMap;
+
+import ic2.core.util.LogCategory;
 import net.minecraft.nbt.NBTTagCompound;
 import net.minecraft.tileentity.TileEntity;
 import net.minecraft.world.World;
@@ -60,6 +62,7 @@
          if (oldWorldData != null) {
             worldData = oldWorldData;
          }
+         IC2.log.debug(LogCategory.EnergyNet, "World with id %d loaded on %s.", world.provider.getDimension(), world.isRemote? "client": "server");
       }
 
        return worldData;
@@ -67,8 +70,9 @@
 
    public static void onWorldUnload(World world) {
       getIndex(!world.isRemote).remove(world.provider.getDimension());
+      IC2.log.debug(LogCategory.EnergyNet, "World with id %d unloaded on %s.", world.provider.getDimension(), world.isRemote? "client": "server");
    }
-
+   
    private static ConcurrentMap<Integer, WorldData> getIndex(boolean simulating) {
       return simulating ? idxServer : idxClient;
    }
