--- a/ic2/core/WorldData.java
+++ b/ic2/core/WorldData.java
@@ -13,6 +13,9 @@
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.ConcurrentLinkedQueue;
 import java.util.concurrent.ConcurrentMap;
+
+import ic2.core.util.LogCategory;
+import mods.su5ed.ic2patcher.IC2PatcherConfig;
 import net.minecraft.nbt.NBTTagCompound;
 import net.minecraft.tileentity.TileEntity;
 import net.minecraft.world.World;
@@ -33,6 +36,7 @@
    public final Map<Chunk, NBTTagCompound> worldGenData = new IdentityHashMap<>();
    public final Set<Chunk> chunksToDecorate = Collections.newSetFromMap(new IdentityHashMap<>());
    public final Set<Chunk> pendingUnloadChunks = Collections.newSetFromMap(new IdentityHashMap<>());
+   private final World world;
 
    private WorldData(World world) {
       if (!world.isRemote) {
@@ -42,34 +46,59 @@
          this.energyNet = null;
          this.windSim = null;
       }
+      this.world = world;
    }
 
    public static WorldData get(World world) {
       return get(world, true);
    }
 
+   private static boolean worldDataWarningDetails = false;
    public static WorldData get(World world, boolean load) {
-      if (world == null) {
-         throw new IllegalArgumentException("world is null");
-      } else {
-         ConcurrentMap<Integer, WorldData> index = getIndex(!world.isRemote);
-         WorldData ret = index.get(world.provider.getDimension());
-         if (ret == null && load) {
+      // Patch - WorldData contains invalid World Reference. Fixes a lot of issues related to EnergyNet.
+      // Experimental: Disable this patch for using DimensionManager for world reference. Appears to work worse, however?
+      if (world == null) throw new IllegalArgumentException("world is null");
+
+      ConcurrentMap<Integer, WorldData> index = getIndex(!world.isRemote);
+      WorldData ret = index.get(world.provider.getDimension());
+      if (ret == null && load) {
+         ret = new WorldData(world);
+         WorldData prev = index.putIfAbsent(world.provider.getDimension(), ret);
+         if (prev != null) {
+            ret = prev;
+         }
+         IC2.log.debug(LogCategory.EnergyNet, "World with id %d loaded on %s.", world.provider.getDimension(), world.isRemote? "client": "server");
+      } else if (!IC2PatcherConfig.get().worldReferenceSwitch && load && world != ret.world) {
+         if (world.getClass().equals(ret.world.getClass())) {
+            World oldWorld = ret.world;
             ret = new WorldData(world);
-            WorldData prev = index.putIfAbsent(world.provider.getDimension(), ret);
-            if (prev != null) {
-               ret = prev;
-            }
-
-            return ret;
+            index.put(world.provider.getDimension(), ret);
+            IC2.log.warn(LogCategory.EnergyNet, "The dimension %d had the wrong world referenced (%s) and was replaced with the current one (%s).", world.provider.getDimension(), oldWorld.toString(), world.toString());
+            IC2.log.warn(LogCategory.EnergyNet, "This is likely caused by missing or incorrect ordered world load / unload events.");
          } else {
-            return ret;
+            IC2.log.error(LogCategory.EnergyNet, "The dimension %d has the wrong world referenced (%s), however new reference (%s) isn't the same class as the original!", ret.world.provider.getDimension(), ret.world.toString(), world.toString());
+            if (!worldDataWarningDetails) {
+               IC2.log.error(LogCategory.EnergyNet, "This might cause issues with your EnergyNet, and counts as incompatibility with mod overriding the world object.");
+               IC2.log.error(LogCategory.EnergyNet, "World reference wasn't replaced, leaving old WorldData, EnergyNet and WindSim in place.");
+               IC2.log.error(LogCategory.EnergyNet, "Expected: Object of %s, got: %s", ret.world.toString(), world.toString());
+               worldDataWarningDetails = true;
+            }
          }
       }
+
+      return ret;
    }
 
    public static void onWorldUnload(World world) {
       getIndex(!world.isRemote).remove(world.provider.getDimension());
+      IC2.log.debug(LogCategory.EnergyNet, "World with id %d unloaded on %s.", world.provider.getDimension(), world.isRemote? "client": "server");
+   }
+
+   public static void resetMaps() {
+      idxServer.clear();
+      if (idxClient != null) {
+         idxClient.clear();
+      }
    }
 
    private static ConcurrentMap<Integer, WorldData> getIndex(boolean simulating) {
