--- a/ic2/core/block/machine/tileentity/TileEntityFluidBottler.java
+++ b/ic2/core/block/machine/tileentity/TileEntityFluidBottler.java
@@ -68,7 +68,7 @@
    }
 
    @Override
-   public MachineRecipeResult<Void, Object, Object> getOutput() {
+   public MachineRecipeResult getOutput() {
       MachineRecipeResult<Void, IEmptyFluidContainerRecipeManager.Output, ItemStack> emptyRes = Recipes.emptyFluidContainer
          .apply(
             this.drainInputSlot.get(),
