--- a/ic2/core/block/machine/tileentity/TileEntityScanner.java
+++ b/ic2/core/block/machine/tileentity/TileEntityScanner.java
@@ -146,10 +146,10 @@
    public void readFromNBT(NBTTagCompound nbttagcompound) {
       super.readFromNBT(nbttagcompound);
       this.progress = nbttagcompound.getInteger("progress");
-      NBTTagCompound nbttagcompound = nbttagcompound.getCompoundTag("currentStack");
-      this.currentStack = new ItemStack(nbttagcompound);
-      nbttagcompound = nbttagcompound.getCompoundTag("pattern");
-      this.pattern = new ItemStack(nbttagcompound);
+      NBTTagCompound tag = nbttagcompound.getCompoundTag("currentStack");
+      this.currentStack = new ItemStack(tag);
+      tag = nbttagcompound.getCompoundTag("pattern");
+      this.pattern = new ItemStack(tag);
       int i = nbttagcompound.getInteger("state");
       this.state = i < TileEntityScanner.State.values().length ? TileEntityScanner.State.values()[i] : TileEntityScanner.State.IDLE;
       this.refreshInfo();
