--- a/ic2/core/block/machine/tileentity/TileEntityCompressor.java
+++ b/ic2/core/block/machine/tileentity/TileEntityCompressor.java
@@ -80,7 +80,7 @@
 
                if (fluid.amount <= 0) {
                   this.usingPumpRecipe = true;
-                  output = new MachineRecipe<>(null, Collections.singletonList(new ItemStack(Items.SNOWBALL))).getResult(null);
+                  output = new MachineRecipe(null, Collections.singletonList(new ItemStack(Items.SNOWBALL))).getResult(null);
                   break;
                }
             }
