--- a/ic2/core/block/machine/gui/GuiMetalFormer.java
+++ b/ic2/core/block/machine/gui/GuiMetalFormer.java
@@ -56,11 +56,12 @@
          }
       }));
       if (RecipeButton.canUse()) {
-         for (final int i = 0; i < 3; i++) {
+         for (int i = 0; i < 3; i++) {
+            int j = i;
             this.addElement(new RecipeButton(this, 52, 39, 46, 9, new String[]{"metal_former" + i}).withEnableHandler(new IEnableHandler() {
                @Override
                public boolean isEnabled() {
-                  return container.base.getMode() == i;
+                  return container.base.getMode() == j;
                }
             }));
          }
