--- a/ic2/core/block/invslot/InvSlotUpgrade.java
+++ b/ic2/core/block/invslot/InvSlotUpgrade.java
@@ -1,16 +1,10 @@
 package ic2.core.block.invslot;
 
-import ic2.api.upgrade.IAugmentationUpgrade;
-import ic2.api.upgrade.IEnergyStorageUpgrade;
-import ic2.api.upgrade.IFullUpgrade;
-import ic2.api.upgrade.IProcessingUpgrade;
-import ic2.api.upgrade.IRedstoneSensitiveUpgrade;
-import ic2.api.upgrade.ITransformerUpgrade;
-import ic2.api.upgrade.IUpgradableBlock;
-import ic2.api.upgrade.IUpgradeItem;
+import ic2.api.upgrade.*;
 import ic2.core.block.TileEntityInventory;
 import ic2.core.block.comp.Redstone;
 import ic2.core.block.comp.TileEntityComponent;
+import ic2.core.block.machine.tileentity.TileEntityElectricMachine;
 import ic2.core.util.StackUtil;
 import java.util.ArrayList;
 import java.util.Collections;
@@ -91,6 +85,11 @@
                }
             }
          }
+      }
+
+      if (block instanceof TileEntityElectricMachine && block.getUpgradableProperties().contains(UpgradableProperty.Transformer)) {
+         TileEntityElectricMachine te = (TileEntityElectricMachine) block;
+         te.dischargeSlot.setTier(te.dischargeSlot.baseTier + extraTier);
       }
 
       for (TileEntityComponent component : this.base.getComponents()) {
