--- ../src-base/minecraft/ic2/core/block/storage/box/TileEntitySteelStorageBox.java
+++ ../src-work/minecraft/ic2/core/block/storage/box/TileEntitySteelStorageBox.java
@@ -1,5 +1,8 @@
 package ic2.core.block.storage.box;
 
+import ic2.core.profile.NotClassic;
+
+@NotClassic
 public class TileEntitySteelStorageBox extends TileEntityStorageBox {
    public TileEntitySteelStorageBox() {
       super(63);
