--- a/ic2/core/crop/TileEntityCrop.java
+++ b/ic2/core/crop/TileEntityCrop.java
@@ -499,8 +499,9 @@
       value += height;
       int fresh = 9;
 
-      for (int x = this.pos.getX() - 1; x < this.pos.getX() + 1 && fresh > 0; x++) {
-         for (int z = this.pos.getZ() - 1; z < this.pos.getZ() + 1 && fresh > 0; z++) {
+      // Patch - Air Quality Calculation checks 2x2 area instead of 3x3. (#32)
+      for (int x = this.pos.getX() - 1; x <= this.pos.getX() + 1 && fresh > 0; x++) {
+         for (int z = this.pos.getZ() - 1; z <= this.pos.getZ() + 1 && fresh > 0; z++) {
             if (world.isBlockNormalCube(new BlockPos(x, this.pos.getY(), z), false)
                || world.getTileEntity(new BlockPos(x, this.pos.getY(), z)) instanceof TileEntityCrop) {
                fresh--;
@@ -535,7 +536,8 @@
       World world = this.getWorld();
       int nutrients = Crops.instance.getNutrientBiomeBonus(BiomeUtil.getBiome(world, this.pos));
 
-      for (int i = 1; i < 5 && world.getBlockState(this.pos.down(i)).getBlock() == Blocks.DIRT; i++) {
+      // Patch: Nutrient bonus is not calculated properly, always failing on the initial Farmland block (#35)
+      for (int i = 2; i < 5 && world.getBlockState(this.pos.down(i)).getBlock() == Blocks.DIRT; i++) {
          nutrients++;
       }
 
