--- a/ic2/core/item/armor/ItemArmorJetpack.java
+++ b/ic2/core/item/armor/ItemArmorJetpack.java
@@ -3,6 +3,7 @@
 import ic2.core.item.armor.jetpack.IJetpack;
 import ic2.core.ref.FluidName;
 import ic2.core.ref.ItemName;
+import ic2.core.util.StackUtil;
 import ic2.core.util.Util;
 import net.minecraft.creativetab.CreativeTabs;
 import net.minecraft.item.ItemStack;
@@ -56,7 +57,7 @@
    }
 
    public boolean isJetpackActive(ItemStack stack) {
-      return true;
+      return !StackUtil.getOrCreateNbtData(stack).getBoolean("jetpackDisabled");
    }
 
    public double getChargeLevel(ItemStack stack) {
