--- ../src-base/minecraft/ic2/core/item/tool/GuiToolMeter.java
+++ ../src-work/minecraft/ic2/core/item/tool/GuiToolMeter.java
@@ -31,11 +31,11 @@
 
    protected void mouseClicked(int i, int j, int k) throws IOException {
       super.mouseClicked(i, j, k);
-      int i = (this.width - this.xSize) / 2;
-      int j = (this.height - this.ySize) / 2;
-      int k = i - i;
-      int l = j - j;
-      if (k >= 26 && l >= 111 && k <= 83 && l <= 123) {
+      int i1 = (this.width - this.xSize) / 2;
+      int j1 = (this.height - this.ySize) / 2;
+      int k1 = i - i1;
+      int l = j - j1;
+      if (k1 >= 26 && l >= 111 && k1 <= 83 && l <= 123) {
          ((ContainerMeter)this.container).reset();
       }
 
