--- a/ic2/core/energy/grid/EnergyNetLocal.java
+++ b/ic2/core/energy/grid/EnergyNetLocal.java
@@ -24,6 +24,7 @@
 import net.minecraft.util.EnumFacing;
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.World;
+import net.minecraftforge.common.DimensionManager;
 
 public class EnergyNetLocal {
    static final GridChange QUEUE_DELAY_CHANGE = new GridChange(null, null, null);
@@ -31,7 +32,7 @@
    private final Map<IEnergyTile, GridChange> gridAdditionsMap = new IdentityHashMap<>();
    private final Set<BlockPos> positionsToNotify = new HashSet<>();
    private final GridUpdater updater = new GridUpdater(this);
-   private final World world;
+   private final int worldId;
    int nextNodeId;
    int nextGridId;
    final Map<IEnergyTile, Tile> registeredIoTiles = new IdentityHashMap<>();
@@ -44,7 +45,7 @@
    }
 
    private EnergyNetLocal(World world) {
-      this.world = world;
+      worldId = world.provider.getDimension();
 
       for(int i = 0; i < 1; ++i) {
          this.gridChangesQueue.add(QUEUE_DELAY_CHANGE);
@@ -189,8 +190,9 @@
             Block block = BlockName.te.getInstance();
 
             for(BlockPos blockpos : this.positionsToNotify) {
-               if (this.world.isBlockLoaded(blockpos)) {
-                  this.world.getBlockState(blockpos).neighborChanged(this.world, blockpos, block, blockpos);
+               World world = DimensionManager.getWorld(worldId);
+               if (world.isBlockLoaded(blockpos)) {
+                  world.getBlockState(blockpos).neighborChanged(world, blockpos, block, blockpos);
                }
             }
 
@@ -217,7 +219,7 @@
    }
 
    public World getWorld() {
-      return this.world;
+      return DimensionManager.getWorld(worldId);
    }
 
    int allocateNodeId() {
