--- a/ic2/core/gui/Ic2GuiFactory.java
+++ b/ic2/core/gui/Ic2GuiFactory.java
@@ -74,11 +74,11 @@
             Iterator<Config> iterator = config.sectionIterator();
 
             while(iterator.hasNext()) {
-               Config config = iterator.next();
-               if ("predefined".equals(config.name) && ".balance.uu-values.".equals(parentName)) {
+               Config section = iterator.next();
+               if ("predefined".equals(section.name) && ".balance.uu-values.".equals(parentName)) {
                   list.add(new Ic2GuiFactory.IC2ConfigGuiScreen.UUListElement());
                } else {
-                  list.add(new DummyCategoryElement(config.name, "ic2.config.sub." + config.name, sinkCategoryLevel(config, parentName + config.name + '.')));
+                  list.add(new DummyCategoryElement(section.name, "ic2.config.sub." + section.name, sinkCategoryLevel(section, parentName + section.name + '.')));
                }
             }
 
