--- a/ic2/core/gui/PlayerHead.java
+++ b/ic2/core/gui/PlayerHead.java
@@ -27,9 +27,7 @@
       super(gui, x, y, IMAGE_MAKER.computeIfAbsent(player, (profile) -> {
          ItemStack itemstack = new ItemStack(Items.SKULL, 1, 3);
          StackUtil.getOrCreateNbtData(itemstack).setTag("SkullOwner", NBTUtil.writeGameProfile(new NBTTagCompound(), profile));
-         return (void)(() -> {
-            return (void)skull;
-         });
+         return () -> itemstack;
       }));
       this.player = player;
 
