--- ../src-base/minecraft/ic2/core/ref/ItemName.java
+++ ../src-work/minecraft/ic2/core/ref/ItemName.java
@@ -241,7 +241,7 @@
          return null;
       } else if (this.instance instanceof IMultiItem) {
          IMultiItem<T> imultiitem = (IMultiItem)this.instance;
-         return imultiitem.getItemStack((IIdProvider)variant);
+         return imultiitem.getItemStack((T) variant);
       } else if (variant == null) {
          return new ItemStack(this.instance);
       } else {
