--- a/ic2/core/network/DataEncoder.java
+++ b/ic2/core/network/DataEncoder.java
@@ -409,7 +409,7 @@
                   assert constants != null;
 
                   for (int i = 0; i < len; i++) {
-                     Array.set(array, i, constants[decode(is, componentType)]);
+                     Array.set(array, i, constants[(int) decode(is, componentType)]);
                   }
                } else {
                   for (int i = 0; i < len; i++) {
