--- a/ic2/core/block/machine/tileentity/TileEntityBatchCrafter.java
+++ b/ic2/core/block/machine/tileentity/TileEntityBatchCrafter.java
@@ -140,24 +140,7 @@
          this.ingredientsRow[slot] = new InvSlot(this, "ingredient[" + slot + ']', InvSlot.Access.I, 1) {
             @Override
             public boolean accepts(ItemStack ingredient) {
-               IRecipe recipe = TileEntityBatchCrafter.this.world.isRemote ? TileEntityBatchCrafter.this.findRecipe() : TileEntityBatchCrafter.this.recipe;
-               if (recipe == null) {
-                  return false;
-               } else {
-                  assert recipe.matches(TileEntityBatchCrafter.this.crafting, TileEntityBatchCrafter.this.world);
-
-                  ItemStack recipeStack = TileEntityBatchCrafter.this.craftingGrid[slot];
-
-                  boolean var4;
-                  try {
-                     TileEntityBatchCrafter.this.craftingGrid[slot] = ingredient;
-                     var4 = recipe.matches(TileEntityBatchCrafter.this.crafting, TileEntityBatchCrafter.this.world);
-                  } finally {
-                     TileEntityBatchCrafter.this.craftingGrid[slot] = recipeStack;
-                  }
-
-                  return var4;
-               }
+               return StackUtil.checkItemEquality(craftingGrid[slot], ingredient);
             }
 
             @Override
@@ -301,7 +284,7 @@
    }
 
    protected void doCrafting() {
-      for (int operation = 0; operation < this.operationsPerTick; operation++) {
+      for (int operation = 0; operation < this.operationsPerTick && canCraft(); operation++) {
          List<ItemStack> outputs = Collections.singletonList(this.recipeOutput);
 
          for (ItemStack stack : this.upgradeSlot) {
