--- a/ic2/core/profile/ProfileManager.java
+++ b/ic2/core/profile/ProfileManager.java
@@ -46,7 +46,7 @@
       map.put("Experimental", new Profile("Experimental", Collections.singleton(TextureStyle.EXPERIMENTAL), Version.NEW, new RecipeChange[0]));
 
       try {
-         map.put("Classic", ProfileParser.parse(ProfileTarget.fromJar("ic2/profiles/classic")));
+         map.put("Classic", ProfileParser.parse(new ProfileTarget(new File("D:\\Github stuff\\IC2-Patcher\\projects\\IC2-Patched\\src\\main\\java"), "ic2/profiles/classic")));
          return map;
       } catch (IOException ioexception) {
          throw new RuntimeException("Error opening profile XML", ioexception);
