--- ../src-base/minecraft/ic2/core/block/machine/gui/GuiFluidRegulator.java
+++ ../src-work/minecraft/ic2/core/block/machine/gui/GuiFluidRegulator.java
@@ -29,48 +29,48 @@
 
    protected void mouseClicked(int i, int j, int k) throws IOException {
       super.mouseClicked(i, j, k);
-      int i = (this.width - this.xSize) / 2;
-      int j = (this.height - this.ySize) / 2;
-      int k = i - i;
-      int l = j - j;
+      int i1 = (this.width - this.xSize) / 2;
+      int j1 = (this.height - this.ySize) / 2;
+      int k1 = i1 - i1;
+      int l = j1 - j1;
       TileEntityFluidRegulator tileentityfluidregulator = (TileEntityFluidRegulator)((ContainerFluidRegulator)this.container).base;
-      if (k >= 102 && l >= 68 && k <= 110 && l <= 76) {
+      if (k1 >= 102 && l >= 68 && k1 <= 110 && l <= 76) {
          ((NetworkManager)IC2.network.get(false)).initiateClientTileEntityEvent(tileentityfluidregulator, -1000);
       }
 
-      if (k >= 112 && l >= 68 && k <= 120 && l <= 76) {
+      if (k1 >= 112 && l >= 68 && k1 <= 120 && l <= 76) {
          ((NetworkManager)IC2.network.get(false)).initiateClientTileEntityEvent(tileentityfluidregulator, -100);
       }
 
-      if (k >= 122 && l >= 68 && k <= 130 && l <= 76) {
+      if (k1 >= 122 && l >= 68 && k1 <= 130 && l <= 76) {
          ((NetworkManager)IC2.network.get(false)).initiateClientTileEntityEvent(tileentityfluidregulator, -10);
       }
 
-      if (k >= 132 && l >= 68 && k <= 140 && l <= 76) {
+      if (k1 >= 132 && l >= 68 && k1 <= 140 && l <= 76) {
          ((NetworkManager)IC2.network.get(false)).initiateClientTileEntityEvent(tileentityfluidregulator, -1);
       }
 
-      if (k >= 132 && l >= 44 && k <= 140 && l <= 52) {
+      if (k1 >= 132 && l >= 44 && k1 <= 140 && l <= 52) {
          ((NetworkManager)IC2.network.get(false)).initiateClientTileEntityEvent(tileentityfluidregulator, 1);
       }
 
-      if (k >= 122 && l >= 44 && k <= 130 && l <= 52) {
+      if (k1 >= 122 && l >= 44 && k1 <= 130 && l <= 52) {
          ((NetworkManager)IC2.network.get(false)).initiateClientTileEntityEvent(tileentityfluidregulator, 10);
       }
 
-      if (k >= 112 && l >= 44 && k <= 120 && l <= 52) {
+      if (k1 >= 112 && l >= 44 && k1 <= 120 && l <= 52) {
          ((NetworkManager)IC2.network.get(false)).initiateClientTileEntityEvent(tileentityfluidregulator, 100);
       }
 
-      if (k >= 102 && l >= 44 && k <= 110 && l <= 52) {
+      if (k1 >= 102 && l >= 44 && k1 <= 110 && l <= 52) {
          ((NetworkManager)IC2.network.get(false)).initiateClientTileEntityEvent(tileentityfluidregulator, 1000);
       }
 
-      if (k >= 151 && l >= 44 && k <= 161 && l <= 52) {
+      if (k1 >= 151 && l >= 44 && k1 <= 161 && l <= 52) {
          ((NetworkManager)IC2.network.get(false)).initiateClientTileEntityEvent(tileentityfluidregulator, 1001);
       }
 
-      if (k >= 151 && l >= 68 && k <= 161 && l <= 76) {
+      if (k1 >= 151 && l >= 68 && k1 <= 161 && l <= 76) {
          ((NetworkManager)IC2.network.get(false)).initiateClientTileEntityEvent(tileentityfluidregulator, 1002);
       }
 
