--- ../src-base/minecraft/ic2/core/energy/EnergyNetLocal.java
+++ ../src-work/minecraft/ic2/core/energy/EnergyNetLocal.java
@@ -417,14 +417,14 @@
          for(IEnergyTile ienergytile : tile.subTiles) {
             for(EnumFacing enumfacing : EnumFacing.VALUES) {
                BlockPos blockpos = EnergyNet.instance.getPos(ienergytile).offset(enumfacing);
-               Tile tile = this.registeredTiles.get(blockpos);
-               if (tile != null && tile != node.tile) {
-                  for(Node node1 : tile.nodes) {
+               Tile registeredTile = this.registeredTiles.get(blockpos);
+               if (registeredTile != null && registeredTile != node.tile) {
+                  for(Node node1 : registeredTile.nodes) {
                      if (!node1.isExtraNode()) {
                         boolean flag = false;
                         if ((node.nodeType == NodeType.Source || node.nodeType == NodeType.Conductor) && node1.nodeType != NodeType.Source) {
                            IEnergyEmitter ienergyemitter = (IEnergyEmitter)(ienergytile instanceof IEnergyEmitter ? ienergytile : node.tile.mainTile);
-                           IEnergyTile ienergytile1 = tile.getSubTileAt(blockpos);
+                           IEnergyTile ienergytile1 = registeredTile.getSubTileAt(blockpos);
                            IEnergyAcceptor ienergyacceptor = (IEnergyAcceptor)(ienergytile1 instanceof IEnergyAcceptor ? ienergytile1 : node1.tile.mainTile);
                            flag = ienergyemitter.emitsEnergyTo((IEnergyAcceptor)node1.tile.mainTile, enumfacing) && ienergyacceptor.acceptsEnergyFrom((IEnergyEmitter)node.tile.mainTile, enumfacing.getOpposite());
                         }
@@ -432,7 +432,7 @@
                         boolean flag3 = false;
                         if (!flag && (node.nodeType == NodeType.Sink || node.nodeType == NodeType.Conductor) && node1.nodeType != NodeType.Sink) {
                            IEnergyAcceptor ienergyacceptor1 = (IEnergyAcceptor)(ienergytile instanceof IEnergyAcceptor ? ienergytile : node.tile.mainTile);
-                           IEnergyTile ienergytile2 = tile.getSubTileAt(blockpos);
+                           IEnergyTile ienergytile2 = registeredTile.getSubTileAt(blockpos);
                            IEnergyEmitter ienergyemitter1 = (IEnergyEmitter)(ienergytile2 instanceof IEnergyEmitter ? ienergytile2 : node1.tile.mainTile);
                            flag3 = ienergyacceptor1.acceptsEnergyFrom((IEnergyEmitter)node1.tile.mainTile, enumfacing) && ienergyemitter1.emitsEnergyTo((IEnergyAcceptor)node.tile.mainTile, enumfacing.getOpposite());
                         }
