--- ../src-base/minecraft/ic2/core/energy/Tile.java
+++ ../src-work/minecraft/ic2/core/energy/Tile.java
@@ -57,9 +57,9 @@
       if (node.isExtraNode()) {
          flag = true;
       } else {
-         for(Node node : this.nodes) {
-            if (node != node && node.nodeType == node.nodeType && node.isExtraNode()) {
-               node.setExtraNode(false);
+         for(Node myNode : this.nodes) {
+            if (myNode != node && myNode.nodeType == node.nodeType && myNode.isExtraNode()) {
+               myNode.setExtraNode(false);
                flag = true;
                break;
             }
