--- a/ic2/core/block/steam/TileEntityCokeKilnHatch.java
+++ b/ic2/core/block/steam/TileEntityCokeKilnHatch.java
@@ -7,6 +7,7 @@
 import ic2.core.gui.dynamic.DynamicContainer;
 import ic2.core.gui.dynamic.DynamicGui;
 import ic2.core.gui.dynamic.GuiParser;
+import ic2.core.profile.NotClassic;
 import net.minecraft.client.gui.GuiScreen;
 import net.minecraft.entity.player.EntityPlayer;
 import net.minecraft.item.ItemStack;
@@ -14,6 +15,7 @@
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
 
+@NotClassic
 public class TileEntityCokeKilnHatch extends TileEntityInventory implements IHasGui {
    protected final InvSlot inventory = new InvSlot(this, "inventory", InvSlot.Access.I, 1, InvSlot.InvSide.ANY);
 
