--- ../src-base/minecraft/ic2/core/uu/DropScan.java
+++ ../src-work/minecraft/ic2/core/uu/DropScan.java
@@ -158,7 +158,7 @@
          try {
             Method method = Thread.class.getDeclaredMethod("getThreads");
             method.setAccessible(true);
-            Thread[] athread = (Thread[])method.invoke((Object)null);
+            Thread[] athread = (Thread[])method.invoke(null);
             Field field = ReflectionUtil.getField(Thread.class, Runnable.class);
 
             for(Thread thread : athread) {
@@ -433,7 +433,7 @@
 
       public void clear() {
          this.extraChunks.clear();
-         Arrays.fill(this.chunks, (Object)null);
+         Arrays.fill(this.chunks, null);
       }
 
       public String makeString() {
