--- a/ic2/core/network/DataEncoder.java
+++ b/ic2/core/network/DataEncoder.java
@@ -619,6 +619,11 @@
          } else if (dst instanceof TileEntityComponent) {
             NBTTagCompound nbt = (NBTTagCompound)src;
             ((TileEntityComponent)dst).readFromNbt(nbt);
+         } else if (dst.getClass().isArray()) {
+            T[] array1 = (T[]) src;
+            T[] array2 = (T[]) dst;
+            if (array1.length != array2.length) return false;
+            System.arraycopy(array1, 0, array2, 0, array1.length);
          } else {
             if (!(dst instanceof Collection)) {
                return false;
