--- ../src-base/minecraft/ic2/core/block/TeBlockRegistry.java
+++ ../src-work/minecraft/ic2/core/block/TeBlockRegistry.java
@@ -212,7 +212,7 @@
       private final List<ITeBlock> idMap;
 
       TeBlockInfo(E universe) {
-         this(universe.getClass());
+         this((Class<E>) universe.getClass());
       }
 
       TeBlockInfo(Class<E> universe) {
@@ -232,7 +232,7 @@
                int i = ((IIdProvider)block).getId();
 
                while(this.idMap.size() < i) {
-                  this.idMap.add((Object)null);
+                  this.idMap.add(null);
                }
 
                if (this.idMap.size() == i) {
