--- a/ic2/core/block/RenderBlockWall.java
+++ b/ic2/core/block/RenderBlockWall.java
@@ -70,15 +70,16 @@
 
       for (EnumFacing side : EnumFacing.VALUES) {
          int[] sideColorMultipliers = colorMultipliers[side.ordinal()];
-         List<BakedQuad> baseFaceQuads = baseModel.getQuads(state, side, 0L);
+         List<BakedQuad> baseFaceQuads = new ArrayList<>(baseModel.getQuads(state, side, 0L));
          if (sideColorMultipliers == null) {
             mergedQuads[side.ordinal()] = baseFaceQuads;
          } else {
             List<BakedQuad> mergedFaceQuads = new ArrayList<>(baseFaceQuads.size() + sideColorMultipliers.length);
-            mergedFaceQuads.addAll(baseFaceQuads);
+            if (sideColorMultipliers.length < 1) mergedFaceQuads.addAll(baseFaceQuads);
+            else baseFaceQuads.clear();
 
             for (int sideColorMultiplier : sideColorMultipliers) {
-               generateQuad(side, 0.001F, buffer);
+               generateQuad(side, 0, buffer);
                mergedFaceQuads.add(BasicBakedBlockModel.createQuad(Arrays.copyOf(buffer.array(), buffer.position()), side, null));
                ((Buffer)buffer).rewind();
             }
