--- a/ic2/core/block/wiring/GuiTransformer.java
+++ b/ic2/core/block/wiring/GuiTransformer.java
@@ -33,9 +33,9 @@
 
    protected void mouseClicked(int i, int j, int k) throws IOException {
       super.mouseClicked(i, j, k);
-      int i = i - this.guiLeft;
-      int j = j - this.guiTop;
-      if (i >= 150 && j >= 32 && i <= 167 && j <= 49) {
+      int i1 = i - this.guiLeft;
+      int j1 = j - this.guiTop;
+      if (i1 >= 150 && j1 >= 32 && i1 <= 167 && j1 <= 49) {
          ((NetworkManager)IC2.network.get(false)).initiateClientTileEntityEvent((TileEntity)((ContainerTransformer)this.container).base, 3);
       }
 
