--- a/ic2/core/profile/Profile.java
+++ b/ic2/core/profile/Profile.java
@@ -29,11 +29,11 @@
          for(RecipeChange recipechange : changes) {
             if (recipechange.type != RecipeChange.ChangeType.REMOVAL) {
                (map.computeIfAbsent(recipechange.name, (k) -> {
-                  return (void)(new ArrayList());
+                  return (new ArrayList());
                })).add(recipechange);
             } else {
                (map1.computeIfAbsent(recipechange.name, (k) -> {
-                  return (void)(new ArrayList());
+                  return (new ArrayList());
                })).add(recipechange);
             }
          }
