--- a/ic2/core/item/armor/ItemArmorNightvisionGoggles.java
+++ b/ic2/core/item/armor/ItemArmorNightvisionGoggles.java
@@ -4,8 +4,11 @@
 import ic2.api.item.IElectricItem;
 import ic2.api.item.IItemHudInfo;
 import ic2.core.IC2;
+import ic2.core.item.BaseElectricItem;
 import ic2.core.item.ElectricItemManager;
+import ic2.core.item.IPseudoDamageItem;
 import ic2.core.ref.ItemName;
+import ic2.core.util.LogCategory;
 import ic2.core.util.StackUtil;
 import java.util.LinkedList;
 import java.util.List;
@@ -20,7 +23,7 @@
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.World;
 
-public class ItemArmorNightvisionGoggles extends ItemArmorUtility implements IElectricItem, IItemHudInfo {
+public class ItemArmorNightvisionGoggles extends ItemArmorUtility implements IElectricItem, IItemHudInfo, IPseudoDamageItem {
    public ItemArmorNightvisionGoggles() {
       super(ItemName.nightvision_goggles, "nightvision", EntityEquipmentSlot.HEAD);
       this.setMaxDamage(27);
@@ -58,6 +61,7 @@
       NBTTagCompound nbtData = StackUtil.getOrCreateNbtData(stack);
       boolean active = nbtData.getBoolean("active");
       byte toggleTimer = nbtData.getByte("toggleTimer");
+      int skylight = player.getEntityWorld().getLightFromNeighbors(new BlockPos(player));
       if (IC2.keyboard.isAltKeyDown(player) && IC2.keyboard.isModeSwitchKeyDown(player) && toggleTimer == 0) {
          toggleTimer = 10;
          active = !active;
@@ -67,6 +71,8 @@
                IC2.platform.messagePlayer(player, "Nightvision enabled.");
             } else {
                IC2.platform.messagePlayer(player, "Nightvision disabled.");
+               if (skylight > 8) IC2.platform.removePotion(player, MobEffects.BLINDNESS);
+               else IC2.platform.removePotion(player, MobEffects.NIGHT_VISION);
             }
          }
       }
@@ -77,7 +83,6 @@
 
       boolean ret = false;
       if (active && IC2.platform.isSimulating() && ElectricItem.manager.use(stack, 1.0, player)) {
-         int skylight = player.getEntityWorld().getLightFromNeighbors(new BlockPos(player));
          if (skylight > 8) {
             IC2.platform.removePotion(player, MobEffects.NIGHT_VISION);
             player.addPotionEffect(new PotionEffect(MobEffects.BLINDNESS, 100, 0, true, true));
@@ -103,5 +108,17 @@
    @Override
    public boolean getIsRepairable(ItemStack par1ItemStack, ItemStack par2ItemStack) {
       return false;
+   }
+
+   public void setDamage(ItemStack stack, int damage) {
+      int i = this.getDamage(stack);
+      if (damage != i && BaseElectricItem.logIncorrectItemDamaging) {
+         IC2.log.warn(LogCategory.Armor, new Throwable(), "Detected invalid armor damage application (%d):", damage - i);
+      }
+   }
+
+   @Override
+   public void setStackDamage(ItemStack var1, int var2) {
+      super.setDamage(var1, var2);
    }
 }
