--- a/ic2/core/audio/AudioManagerClient.java
+++ b/ic2/core/audio/AudioManagerClient.java
@@ -89,10 +89,10 @@
    @SubscribeEvent
    public void onSoundSetup(SoundLoadEvent event) {
       if (this.enabled) {
-         Iterator thread = this.objectToAudioSourceMap.values().iterator();
+         Iterator<List<AudioSourceClient>> it = this.objectToAudioSourceMap.values().iterator();
 
-         while(thread.hasNext()) {
-            for(AudioSourceClient audiosourceclient : (List)thread.next()) {
+         while(it.hasNext()) {
+            for(AudioSourceClient audiosourceclient : it.next()) {
                if (audiosourceclient.isValid()) {
                   audiosourceclient.setInvalid();
                }
