--- a/ic2/core/block/transport/items/PipeModel.java
+++ b/ic2/core/block/transport/items/PipeModel.java
@@ -58,7 +58,7 @@
 
    private static Map<ResourceLocation, TextureAtlasSprite> generateTextureLocations() {
       Map<ResourceLocation, TextureAtlasSprite> map = new HashMap<ResourceLocation, TextureAtlasSprite>();
-      map.put(new ResourceLocation("ic2", "blocks/transport/pipe_side"), (Object)null);
+      map.put(new ResourceLocation("ic2", "blocks/transport/pipe_side"), null);
       StringBuilder stringbuilder = new StringBuilder();
       stringbuilder.append("blocks/transport/");
       stringbuilder.append("pipe");
@@ -67,7 +67,7 @@
       for(PipeSize pipesize : PipeSize.values()) {
          stringbuilder.append('_');
          stringbuilder.append(pipesize.name());
-         map.put(new ResourceLocation("ic2", stringbuilder.toString()), (Object)null);
+         map.put(new ResourceLocation("ic2", stringbuilder.toString()), null);
          stringbuilder.setLength(i);
       }
 
