--- a/ic2/core/item/tool/ItemToolPainter.java
+++ b/ic2/core/item/tool/ItemToolPainter.java
@@ -127,12 +127,12 @@
    }
 
    private boolean colorBlock(World world, BlockPos pos, Block block, IBlockState state, EnumDyeColor newColor) {
-      UnmodifiableIterator var6 = state.getProperties().keySet().iterator();
+      UnmodifiableIterator<IProperty<?>> it = state.getProperties().keySet().iterator();
 
-      while (var6.hasNext()) {
-         IProperty<?> property = (IProperty<?>)var6.next();
+      while (it.hasNext()) {
+         IProperty<EnumDyeColor> property = (IProperty<EnumDyeColor>) it.next();
          if (property.getValueClass() == EnumDyeColor.class) {
-            EnumDyeColor oldColor = (EnumDyeColor)state.getValue(property);
+            EnumDyeColor oldColor = state.getValue(property);
             if (oldColor != newColor && property.getAllowedValues().contains(newColor)) {
                world.setBlockState(pos, state.withProperty(property, newColor));
                return true;
