--- a/ic2/core/item/tool/ItemToolPainter.java
+++ b/ic2/core/item/tool/ItemToolPainter.java
@@ -130,9 +130,9 @@
       UnmodifiableIterator unmodifiableiterator = state.getProperties().keySet().iterator();
 
       while(unmodifiableiterator.hasNext()) {
-         IProperty<?> iproperty = (IProperty)unmodifiableiterator.next();
+         IProperty<EnumDyeColor> iproperty = (IProperty)unmodifiableiterator.next();
          if (iproperty.getValueClass() == EnumDyeColor.class) {
-            EnumDyeColor enumdyecolor = (EnumDyeColor)state.getValue(iproperty);
+            EnumDyeColor enumdyecolor = state.getValue(iproperty);
             if (enumdyecolor != newColor && iproperty.getAllowedValues().contains(newColor)) {
                world.setBlockState(pos, state.withProperty(iproperty, newColor));
                return true;
