--- a/ic2/core/WorldData.java
+++ b/ic2/core/WorldData.java
@@ -2,6 +2,7 @@
 
 import ic2.core.energy.grid.EnergyNetLocal;
 import ic2.core.network.TeUpdateDataServer;
+import ic2.core.util.LogCategory;
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.HashSet;
@@ -22,6 +23,7 @@
 public class WorldData {
    private static ConcurrentMap<Integer, WorldData> idxClient = FMLCommonHandler.instance().getSide().isClient() ? new ConcurrentHashMap() : null;
    private static ConcurrentMap<Integer, WorldData> idxServer = new ConcurrentHashMap<Integer, WorldData>();
+   private final World world;
    final Queue<IWorldTickCallback> singleUpdates = new ConcurrentLinkedQueue<IWorldTickCallback>();
    final Set<IWorldTickCallback> continuousUpdates = new HashSet<IWorldTickCallback>();
    boolean continuousUpdatesInUse = false;
@@ -42,13 +44,14 @@
          this.energyNet = null;
          this.windSim = null;
       }
-
+      this.world = world;
    }
 
    public static WorldData get(World world) {
       return get(world, true);
    }
 
+   private static boolean worldDataWarningDetails = true;
    public static WorldData get(World world, boolean load) {
       if (world == null) {
          throw new IllegalArgumentException("world is null");
@@ -61,9 +64,27 @@
             if (worlddata1 != null) {
                worlddata = worlddata1;
             }
+            IC2.log.debug(LogCategory.EnergyNet, "World with id %d loaded.", world.provider.getDimension());
 
             return worlddata;
          } else {
+            if (load && world != worlddata.world) {
+               if (world.getClass().equals(worlddata.world.getClass())) {
+                  World oldWorld = worlddata.world;
+                  worlddata = new WorldData(world);
+                  concurrentmap.put(world.provider.getDimension(), worlddata);
+                  IC2.log.warn(LogCategory.EnergyNet, "The dimension %d had the wrong world referenced (%s) and was replaced with the current one (%s).", world.provider.getDimension(), oldWorld.toString(), world.toString());
+                  IC2.log.warn(LogCategory.EnergyNet, "This is likely caused by missing or incorrect ordered world load / unload events.");
+               } else {
+                  IC2.log.error(LogCategory.EnergyNet, "The dimension %d has the wrong world referenced (%s), however new reference isn't the same class as the original (%s)!", worlddata.world.provider.getDimension(), worlddata.world.toString(), world.toString());
+                  if (worldDataWarningDetails) {
+                     IC2.log.error(LogCategory.EnergyNet, "This might cause issues with your EnergyNet, and counts as incompatibility with mod overriding the world object.");
+                     IC2.log.error(LogCategory.EnergyNet, "World reference wasn't replaced, leaving old WorldData, EnergyNet and WindSim in place.");
+                     IC2.log.error(LogCategory.EnergyNet, "Expected: Object of %s, got: %s", worlddata.world.toString(), world.toString());
+                     worldDataWarningDetails = true;
+                  }
+               }
+            }
             return worlddata;
          }
       }
@@ -71,6 +92,14 @@
 
    public static void onWorldUnload(World world) {
       getIndex(!world.isRemote).remove(Integer.valueOf(world.provider.getDimension()));
+      IC2.log.debug(LogCategory.EnergyNet, "World with id %d unloaded.", world.provider.getDimension());
+   }
+
+   public static void resetMaps() {
+      idxServer.clear();
+      if (idxClient != null) {
+         idxClient.clear();
+      }
    }
 
    private static ConcurrentMap<Integer, WorldData> getIndex(boolean simulating) {
