--- ../src-base/minecraft/ic2/jeiIntegration/recipe/crafting/AdvShapelessRecipeWrapper.java
+++ ../src-work/minecraft/ic2/jeiIntegration/recipe/crafting/AdvShapelessRecipeWrapper.java
@@ -22,23 +22,15 @@
    }
 
    public List<List<ItemStack>> getInputs() {
-      List<List<ItemStack>> list = new ArrayList<List<ItemStack>>(this.recipe.input.length);
-
-      for(IRecipeInput irecipeinput : this.recipe.input) {
-         list.add(irecipeinput.getInputs());
+      ItemStack stack;
+      ArrayList<List<ItemStack>> ret = new ArrayList<List<ItemStack>>(this.recipe.input.length);
+      for (IRecipeInput input : this.recipe.input) {
+         ret.add(input.getInputs());
       }
-
-      if (list.size() == 1 && ((List)list.get(0)).size() == 1) {
-         ItemStack itemstack = (ItemStack)(list.get(0)).get(0);
-         if (itemstack.getItem() == ItemName.painter.getInstance() && itemstack.getMetadata() == 32767) {
-            Stream stream = Arrays.stream(Ic2Color.values);
-            ItemName itemname = ItemName.painter;
-            ItemName.painter.getClass();
-            list.set(0, stream.map(itemname::getItemStack).collect(Collectors.toList()));
-         }
+      if (ret.size() == 1 && ((List)ret.get(0)).size() == 1 && (stack = (ItemStack)((List)ret.get(0)).get(0)).getItem() == ItemName.painter.getInstance() && stack.getMetadata() == Short.MAX_VALUE) {
+         ret.set(0, Arrays.stream(Ic2Color.values).map(ItemName.painter::getItemStack).collect(Collectors.toList()));
       }
-
-      return list;
+      return ret;
    }
 
    public void getIngredients(IIngredients ingredients) {
