--- ../src-base/minecraft/ic2/core/block/invslot/InvSlotUpgrade.java
+++ ../src-work/minecraft/ic2/core/block/invslot/InvSlotUpgrade.java
@@ -32,7 +32,7 @@
    private List<Redstone.IRedstoneModifier> redstoneModifiers = Collections.<Redstone.IRedstoneModifier>emptyList();
 
    public static InvSlotUpgrade createUnchecked(IInventorySlotHolder<?> base, String name, int count) {
-      return new InvSlotUpgrade((IInventorySlotHolder)((IUpgradableBlock)base), name, count);
+      return new InvSlotUpgrade((IInventorySlotHolder<?> & IUpgradableBlock)base, name, count);
    }
 
    public <T extends IInventorySlotHolder<?> & IUpgradableBlock> InvSlotUpgrade(T base, String name, int count) {
