--- a/ic2/core/WindSim.java
+++ b/ic2/core/WindSim.java
@@ -1,6 +1,8 @@
 package ic2.core;
 
+import mods.su5ed.ic2patcher.IC2PatcherConfig;
 import net.minecraft.world.World;
+import net.minecraftforge.common.DimensionManager;
 import org.ejml.simple.SimpleMatrix;
 
 public class WindSim {
@@ -8,10 +10,12 @@
    private int windDirection = IC2.random.nextInt(360);
    public int windTicker;
    private final World world;
+   private final int worldId;
    private final SimpleMatrix windHeightCoefficients;
 
    public WindSim(World world) {
       this.world = world;
+      this.worldId = world.provider.getDimension();
       this.windHeightCoefficients = calculateCoefficients(IC2.getWorldHeight(world), IC2.getSeaLevel(world));
    }
 
@@ -72,9 +76,10 @@
       x.setRow(0, 0, new double[]{height, height * height, height * height * height});
       double heightMultiplier = Math.max(0.0, x.mult(this.windHeightCoefficients).get(0));
       ret *= heightMultiplier;
-      if (this.world.isThundering()) {
+      World localWorld = IC2PatcherConfig.get().worldReferenceSwitch? DimensionManager.getWorld(worldId): world;
+      if (localWorld.isThundering()) {
          ret *= 1.5;
-      } else if (this.world.isRaining()) {
+      } else if (localWorld.isRaining()) {
          ret *= 1.25;
       }
 
