--- a/ic2/core/block/wiring/CableModel.java
+++ b/ic2/core/block/wiring/CableModel.java
@@ -81,13 +81,13 @@
 
                for(Ic2Color ic2color : Ic2Color.values) {
                   stringbuilder.append(ic2color.name());
-                  map.put(new ResourceLocation("ic2", stringbuilder.toString()), (Object)null);
+                  map.put(new ResourceLocation("ic2", stringbuilder.toString()), null);
                   stringbuilder.setLength(l);
                }
             } else {
-               map.put(new ResourceLocation("ic2", stringbuilder.toString()), (Object)null);
+               map.put(new ResourceLocation("ic2", stringbuilder.toString()), null);
                if (cabletype == CableType.splitter || cabletype == CableType.detector) {
-                  map.put(new ResourceLocation("ic2", stringbuilder.toString() + "_active"), (Object)null);
+                  map.put(new ResourceLocation("ic2", stringbuilder.toString() + "_active"), null);
                }
             }
 
