--- a/ic2/core/gui/dynamic/DynamicHandHeldContainer.java
+++ b/ic2/core/gui/dynamic/DynamicHandHeldContainer.java
@@ -25,15 +25,15 @@
       super.onContainerEvent(event);
    }
 
-   public ItemStack slotClick(int slot, int button, ClickType type, EntityPlayer player) {
+   public ItemStack slotClick(int slotIndex, int button, ClickType type, EntityPlayer player) {
       boolean flag = false;
       Slot slot = null;
-      if (!player.getEntityWorld().isRemote && slot >= 0 && slot < this.inventorySlots.size()) {
-         slot = (Slot)this.inventorySlots.get(slot);
+      if (!player.getEntityWorld().isRemote && slotIndex >= 0 && slotIndex < this.inventorySlots.size()) {
+         slot = (Slot)this.inventorySlots.get(slotIndex);
          flag = ((HandHeldInventory)this.base).isThisContainer(slot.getStack());
       }
 
-      ItemStack itemstack = super.slotClick(slot, button, type, player);
+      ItemStack itemstack = super.slotClick(slotIndex, button, type, player);
       if (flag && !slot.getHasStack()) {
          ((HandHeldInventory)this.base).saveAndThrow(itemstack);
          player.closeScreen();
