--- ../src-base/minecraft/ic2/core/block/RenderBlockWall.java
+++ ../src-work/minecraft/ic2/core/block/RenderBlockWall.java
@@ -68,15 +68,16 @@
 
       for(EnumFacing enumfacing : EnumFacing.VALUES) {
          int[] aint1 = colorMultipliers[enumfacing.ordinal()];
-         List<BakedQuad> list1 = baseModel.getQuads(state, enumfacing, 0L);
+         List<BakedQuad> list1 = new ArrayList<>(baseModel.getQuads(state, enumfacing, 0L));
          if (aint1 == null) {
             list[enumfacing.ordinal()] = list1;
          } else {
             List<BakedQuad> list2 = new ArrayList<BakedQuad>(list1.size() + aint1.length);
-            list2.addAll(list1);
+            if (aint1.length < 1) list2.addAll(list1);
+            else list1.clear();
 
             for(int i : aint1) {
-               generateQuad(enumfacing, 0.001F, intbuffer);
+               generateQuad(enumfacing, 0, intbuffer);
                list2.add(BasicBakedBlockModel.createQuad(Arrays.copyOf(intbuffer.array(), intbuffer.position()), enumfacing, (TextureAtlasSprite)null));
                intbuffer.rewind();
             }
