--- a/ic2/core/util/StackUtil.java
+++ b/ic2/core/util/StackUtil.java
@@ -98,7 +98,7 @@
 
    public static StackUtil.AdjacentInv getAdjacentInventory(TileEntity source, EnumFacing dir) {
       TileEntity tileentity = source.getWorld().getTileEntity(source.getPos().offset(dir));
-      if (!isInventoryTile(tileentity, dir)) {
+      if (!isInventoryTile(tileentity, dir.getOpposite())) {
          return null;
       } else {
          GameProfile gameprofile;
