--- a/ic2/core/item/block/ItemCable.java
+++ b/ic2/core/item/block/ItemCable.java
@@ -223,7 +223,7 @@
                      Ic2Color ic2color = itemtoolpainter.getColor(itemstack1);
                      if (ic2color != null) {
                         runnable = () -> {
-                           painter.damagePainter(player, EnumHand.OFF_HAND, color);
+                           itemtoolpainter.damagePainter(player, EnumHand.OFF_HAND, ic2color);
                         };
                         tileentitycable = TileEntityCable.delegate(cabletype, i, ic2color);
                         break label33;
