--- a/ic2/core/item/armor/ItemArmorQuantumSuit.java
+++ b/ic2/core/item/armor/ItemArmorQuantumSuit.java
@@ -16,6 +16,8 @@
 import java.util.IdentityHashMap;
 import java.util.LinkedList;
 import java.util.Map;
+
+import mods.su5ed.ic2patcher.IC2PatcherConfig;
 import net.minecraft.entity.EntityLivingBase;
 import net.minecraft.entity.player.EntityPlayer;
 import net.minecraft.init.MobEffects;
@@ -229,6 +231,10 @@
 
             boolean Nightvision = nbtData.getBoolean("Nightvision");
             short hubmode = nbtData.getShort("HudMode");
+            int skylight = IC2.platform.isSimulating()? player.getEntityWorld().getLightFromNeighbors(
+                new BlockPos((int)Math.floor(player.posX), (int)Math.floor(player.posY), (int)Math.floor(player.posZ))
+            ): 0;
+
             if (IC2.keyboard.isAltKeyDown(player) && IC2.keyboard.isModeSwitchKeyDown(player) && toggleTimer == 0) {
                toggleTimer = 10;
                Nightvision = !Nightvision;
@@ -238,6 +244,9 @@
                      IC2.platform.messagePlayer(player, "Nightvision enabled.");
                   } else {
                      IC2.platform.messagePlayer(player, "Nightvision disabled.");
+                     // PATCH: Make Night Vision turn off instantly.
+                     if (skylight > 8) IC2.platform.removePotion(player, MobEffects.BLINDNESS);
+                     else IC2.platform.removePotion(player, MobEffects.NIGHT_VISION);
                   }
                }
             }
@@ -261,13 +270,14 @@
             }
 
             if (Nightvision && IC2.platform.isSimulating() && ElectricItem.manager.use(stack, 1.0, player)) {
-               BlockPos pos = new BlockPos((int)Math.floor(player.posX), (int)Math.floor(player.posY), (int)Math.floor(player.posZ));
-               int skylight = player.getEntityWorld().getLightFromNeighbors(pos);
+               // BUFF: Make QuantumSuit Night Vision fully automatic, not applying the blindness if skylight is too high.
                if (skylight > 8) {
                   IC2.platform.removePotion(player, MobEffects.NIGHT_VISION);
-                  player.addPotionEffect(new PotionEffect(MobEffects.BLINDNESS, 100, 0, true, true));
+                  if (!IC2PatcherConfig.get().qsNVBuff)
+                     player.addPotionEffect(new PotionEffect(MobEffects.BLINDNESS, 100, 0, true, true));
                } else {
-                  IC2.platform.removePotion(player, MobEffects.BLINDNESS);
+                  if (!IC2PatcherConfig.get().qsNVBuff)
+                     IC2.platform.removePotion(player, MobEffects.BLINDNESS);
                   player.addPotionEffect(new PotionEffect(MobEffects.NIGHT_VISION, 300, 0, true, true));
                }
 
@@ -384,7 +394,7 @@
 
    @Override
    public boolean isJetpackActive(ItemStack stack) {
-      return true;
+      return !StackUtil.getOrCreateNbtData(stack).getBoolean("jetpackDisabled");
    }
 
    @Override
