--- a/ic2/core/block/machine/FillFluidContainerRecipeManager.java
+++ b/ic2/core/block/machine/FillFluidContainerRecipeManager.java
@@ -39,7 +39,7 @@
                FluidStack changedFluid = result.fluidChange.amount >= input.fluid.amount
                   ? null
                   : new FluidStack(input.fluid, input.fluid.amount - result.fluidChange.amount);
-               return new MachineRecipe<>(null, output).getResult(new IFillFluidContainerRecipeManager.Input(result.inPlaceOutput, changedFluid));
+               return new MachineRecipe(null, output).getResult(new IFillFluidContainerRecipeManager.Input(result.inPlaceOutput, changedFluid));
             }
          }
       } else if (!acceptTest) {
@@ -49,7 +49,7 @@
       } else {
          return !StackUtil.isEmpty(input.container) && !LiquidUtil.isFillableFluidContainer(input.container)
             ? null
-            : new MachineRecipe<>(null, Collections.emptyList()).getResult(input);
+            : new MachineRecipe(null, Collections.emptyList()).getResult(input);
       }
    }
 
