--- a/ic2/core/item/armor/ItemArmorElectric.java
+++ b/ic2/core/item/armor/ItemArmorElectric.java
@@ -29,7 +29,7 @@
    protected final int tier;
 
    public ItemArmorElectric(ItemName name, String armorName, EntityEquipmentSlot armorType, double maxCharge, double transferLimit, int tier) {
-      super(name, ArmorMaterial.DIAMOND, armorName, armorType, (Object)null);
+      super(name, ArmorMaterial.DIAMOND, armorName, armorType, null);
       this.maxCharge = maxCharge;
       this.tier = tier;
       this.transferLimit = transferLimit;
