--- a/ic2/core/block/transport/TileEntityFluidPipe.java
+++ b/ic2/core/block/transport/TileEntityFluidPipe.java
@@ -438,6 +438,7 @@
       }
 
       public IFluidTankProperties[] getTankProperties() {
+         if (TileEntityFluidPipe.this.tank == null) return new IFluidTankProperties[0];
          return TileEntityFluidPipe.this.tank.getTankProperties();
       }
 
@@ -468,7 +469,8 @@
       public FluidStack drain(int maxDrain, boolean doDrain) {
          if (TileEntityFluidPipe.this.coversComponent.hasCover(this.side)) {
             ICoverItem cover = TileEntityFluidPipe.this.coversComponent.getCoverItem(this.side);
-            if (!cover.allowsOutput(new FluidStack(TileEntityFluidPipe.this.tank.getFluid(), maxDrain))) {
+            FluidStack fluid = TileEntityFluidPipe.this.tank.getFluid();
+            if (fluid != null && !cover.allowsOutput(new FluidStack(fluid, maxDrain))) {
                return null;
             }
          }
